### PR TITLE
implemented sample ui interaction tracker event and replace fs read a…

### PR DESCRIPTION
**Pull request by xavluiz**
…nd write

this request contains logic to send a ui_interaction event containing a ui_element entity. The "KpmItem" already has most of the attributes for this, I just added icon and color.

it also swaps out the fs read and write logic in lieu of the file-it NPM to handle all of the try/catch and parse cleanup heavy lifting

### CODE REVIEW

File: src/managers/TrackerManager.ts [hunk 1]
```diff
@@ -27,11 +28,60 @@ export class TrackerManager {
...
+		this.trackUIInteraction("click", ui_element);
+	}
+
+	/**
+	 * @param type execute_command | click
+	 * @param ui_element {element_name, element_location, color, icon_name, cta_text}
+	 */
+	private async trackUIInteraction(interaction_type: string, ui_element: UIElement) {
+		const jwt = getItem("jwt");
+		if (!this.trackerReady || !jwt) {
```
Comment (mmclead) [inline src/managers/TrackerManager.ts:45]: could use getBaseTrackerInfo() and check the jwt on that object.

File: src/managers/TrackerManager.ts [hunk 2]
```diff
@@ -27,11 +28,60 @@ export class TrackerManager {
...
+		}
+
+		const e = {
+			interaction_type,
+			...ui_element,
+			...this.getBaseTrackerInfo()
+		};
+
+		swdcTracker.trackEditorAction(e).then(result => {
+			console.log(`sent editor action: ${JSON.stringify(e)}`)
```
Comment (mmclead) [inline src/managers/TrackerManager.ts:56]: can remove console.log

File: src/command-helper.ts [hunk 3]
```diff
@@ -234,7 +236,11 @@ export function createCommands(
 
     // DISPLAY CODE TIME METRICS REPORT
     cmds.push(
-        commands.registerCommand("codetime.codeTimeMetrics", () => {
+        commands.registerCommand("codetime.codeTimeMetrics", (item: KpmItem) => {
+            if (item) {
+                const uiEl: UIElement = UIElement.transformKpmItemToUIElement(item);
+                TrackerManager.getInstance().trackUICommandInteraction(uiEl);
```
Comment (mmclead) [inline src/command-helper.ts:242]: is this a thing you click on in the TreeView or running the command from the palette?

from what I'm reading the KpmItem is a menu Item in the tree view.  
Can we tell which are clicks vs commands?

File: src/Constants.ts [hunk 4]
```diff
@@ -11,10 +11,10 @@ export const CODE_TIME_TYPE = "codetime";
 
 // API ENDPOINT
 //
-// "http://localhost:5000", "https://qaapi.software.com", "https://stagingapi.software.com", "https://api.software.com"
+// "http://localhost:5000", "https://stagingapi.software.com", "https://api.software.com"
```
[the remark this answers is not in the document]
Comment (xavluiz) [reply src/Constants.ts:14]: just convenience in changing env endpoints quickly for testing but it's not really useful in the beginning, i'll remove it

File: src/tree/KpmProviderManager.ts [hunk 5]
```diff
@@ -555,6 +556,8 @@ export class KpmProviderManager {
         item.icon = icon;
         item.contextValue = "action_button";
         item.eventDescription = eventDescription;
+        item.location = "tree"
```
[the remark this answers is not in the document]
Comment (xavluiz) [reply src/tree/KpmProviderManager.ts:559]: yeah, i'm not tied to "tree", menu-tree is better. or should we specify "ct-menu-tree" for code time?

File: src/managers/TrackerManager.ts (hunk 1 above)
Comment (xavluiz) [reply src/managers/TrackerManager.ts:45]: yeah, thats true. build the object and check from that built object. it should never be null/undefined, but still check on that before sending the event, right?

File: src/command-helper.ts (hunk 3 above)
Comment (xavluiz) [reply src/command-helper.ts:242]: they're all clicks that end up in a registered command. but i see what you're saying. the commands you're thinking of comes from the command palette? both end up in the command-helper and yes, we can determine which is which. i'll update

**Comment by xavluiz** [comment]
Updated the diff based on all of the suggestions.
- New logic includes creating a KpmItem if it comes from the command palette. The registered command will get an undefined KpmItem but we can use the existing button builder to generate that object then replace the location value.

**Comment by xavluiz** [comment]
I'm adding the element name links for the review, thanks.

[Code Time Daily Metrics Tree Elements - location: "ct_metrics_tree"](https://www.notion.so/5a1291a08c7b491b8128c924fdfde950)

[Code Time Contributors Tree Elements - location: "ct_contributors_tree"](https://www.notion.so/fb875fddc7234e19b02744fc60f7b114)

[Code Time Status Bar - location: "ct_status_bar"](https://www.notion.so/f7d4ee5731794dde9924f74c3e56535a)

[Code Time Command Palette - location: "ct_command_palette"](https://www.notion.so/619e0b2faab24480a396566a7c2f9f34)

File: src/managers/TrackerManager.ts [hunk 6]
```diff
@@ -40,24 +98,15 @@ export class TrackerManager {
 		const project_name = (workspaceFolders.length) ? workspaceFolders[0].name : "";
 
 		const token = jwt.split("JWT ")[1];
```
Comment (bjacobson26) [inline src/managers/TrackerManager.ts:100]: is there any chance the jwt will be null? might want to protect against `null.split`

File: src/tree/KpmProviderManager.ts [hunk 7]
```diff
@@ -82,6 +82,7 @@ export class KpmProviderManager {
         const authType: string = getItem("authType");
 
         if (!loggedIn) {
+            // UNCOMMENT when we're ready to provide this feature
```
Comment (bjacobson26) [inline src/tree/KpmProviderManager.ts:85]: should we use a feature flag instead? 

`if(process.env.LINK_TO_ACCOUNT_ENABLED === 'true')`

Comment (xavluiz) [reply src/tree/KpmProviderManager.ts]: good call, added it

File: src/managers/TrackerManager.ts (hunk 6 above)
Comment (xavluiz) [reply src/managers/TrackerManager.ts]: yeah, it was higher up before but that was changed. good catch, thank you. updated

**Comment by xavluiz** [comment]
The latest diff has the following clicks and keyboard events tracked:
ct_summary_btn, ct_project_summary_btn, ct_learn_more_btn, ct_hide_status_bar_metrics_btn, ct_web_metrics_btn, ct_sign_up_google_btn, ct_sign_up_github_btn, ct_sign_up_email_btn

**Pull request merged**